include/sys/onoff.h

```diff
@@ -0,0 +1,603 @@
+	 *
+	 * See async_operation_init_spinwait().
+	 */
+	ONOFF_CLIENT_NOTIFY_SPINWAIT    = 1,
+
+	/*
+	 * Select notification through @ref k_poll signal
+	 *
+	 * See onoff_client_init_signal().
+	 */
+	ONOFF_CLIENT_NOTIFY_SIGNAL      = 2,
+
+	/**
+	 * Select notification through a user-provided callback.
+	 *
+	 * See onoff_client_init_callback().
+	 */
+	ONOFF_CLIENT_NOTIFY_CALLBACK    = 3,
+};
+
+/* Forward declaration */
+struct onoff_client;
+
+/**
+ * @brief Signature used to notify an on-off service client of the
+ * completion of an operation.
+ *
+ * These functions may be invoked from any context including
+ * pre-kernel, ISR, or cooperative or pre-emptible threads.
+ * Compatible functions must be isr-callable and non-suspendable.
+ *
+ * @param srv the service for which the operation was initiated.
+ *
+ * @param cli the client structure passed to the function that
+ * initiated the operation.
+ *
+ * @param user_data user data provided when the client structure was
+ * initialized with onoff_client_init_callback().
+ *
+ * @param res the result of the operation.  Expected values are
+ * service-specific, but the value shall be non-negative if the
+ * operation succeeded, and negative if the operation failed.
+ */
+typedef void (*onoff_client_callback)(struct onoff_service *srv,
+				      struct onoff_client *cli,
+				      void *user_data,
+				      int res);
+
+/**
+ * @brief State associated with a client of an on-off service.
+ *
+ * Objects of this type are allocated by a client, which must use an
+ * initialization function (e.g. onoff_client_init_signal()) to
+ * configure them.
+ *
+ * Control of the object content transfers to the service provider
+ * when a pointer to the object is passed to any on-off service
+ * function.  While the service provider controls the object the
+ * client must not change any object fields.  Control reverts to the
+ * client:
+ * * if the call to the service API returns an error;
+ * * if the call to the service API succeeds for a no-wait operation;
+ * * when operation completion is posted (signalled or callback
+ *   invoked).
+ *
+ * Only the result field is intended for direct use by clients, and it
+ * is available for inspection only after control reverts to the
+ * client.
+ */
+struct onoff_client {
+	/* Links the client into the set of waiting service users. */
+	sys_snode_t node;
+
+	union async {
+		/* Pointer to signal used to notify client.
+		 *
+		 * The signal value corresponds to the res parameter
+		 * of onoff_client_callback.
+		 */
+		struct k_poll_signal *signal;
+
+		/* Handler and argument for callback notification. */
+		struct callback {
+			onoff_client_callback handler;
+			void *user_data;
+		} callback;
+	} async;
+
+	/*
+	 * The result of the operation.
+	 *
+	 * This is the value that was (or would be) passed to the
+	 * async infrastructure.  This field is the sole record of
+	 * success or failure for no-wait synchronous operations.
+	 */
+	int volatile result;
```

Comment:
result is used only for spinwait mode when reading `onoff_client_fetch_result()`. Should be moved to async structure.

Reply:
`onoff_client_fetch_result()` is not specific to spinwait.  It's useful (albeit not necessary) for callbacks, and useful and necessary for some use of signals as well.  For example, you can start multiple services that all notify through the same signal; the result stored in the signal will be the last one received, while the individual results can be obtained from the client structure.

Reply:
> For example, you can start multiple services that all notify through the same signal

Can you do that? How you can synchronize when multiple source can set the signal? It might be that i just don't understand the use case. Anyway, if you see case for having it for other modes then i'm ok with it.

Reply:
Sure; the client structure doesn't own the signal, it just uses one that the application provides.  It's easy to share a signal and when it's raised look at other state (here, the result in the client structure) to determine what's ready.  Saves memory to only require one signal structure.